core/node/metadata_calculator/src/helpers.rs

```diff
@@ -552,10 +553,34 @@ impl L1BatchWithLogs {
         };
 
         let load_tree_writes_latency = METRICS.start_load_stage(LoadChangesStage::LoadTreeWrites);
-        let tree_writes = storage
-            .blocks_dal()
-            .get_tree_writes(l1_batch_number)
-            .await?;
+
+        let tree_writes = if wait_for_tree_writes {
```

Comment:
This flag certainly doesn't feel like a right approach. Intuitively, I'd try logic like this:

- Try getting tree writes immediately. If they are there, use them.
- Check whether tree writes are present for the previous batch (a potential alternative: look at last N batches, N > 1 but still small so that we don't need to do anything resembling a full scan of `l1_batches`). If they are there, wait for writes of the current batch; if not, use a fallback.

So it's similar to what you've implemented *in practice*, but IMO significantly more tractable because of locality; we're not given a parameter from outside whether to wait for tree writes, but decide on our own given reasoning that can ostensibly hold regardless of how tree writes generation is handled.